### PR TITLE
8301492: Modernize equals() method of ResourceBundle.CacheKey and Bundles.CacheKey

### PR DESCRIPTION
**Pull request by stsypanov**
`ResourceBundle.CacheKey.equals()` and `Bundles.CacheKey.equals()` are quire outdated. This simple clean-up modernizes them.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301492](https://bugs.openjdk.org/browse/JDK-8301492): Modernize equals() method of ResourceBundle.CacheKey and Bundles.CacheKey (**Enhancement** - P4)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12328/head:pull/12328` \
`$ git checkout pull/12328`

Update a local copy of the PR: \
`$ git checkout pull/12328` \
`$ git pull https://git.openjdk.org/jdk.git pull/12328/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12328`

View PR using the GUI difftool: \
`$ git pr show -t 12328`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12328.diff">https://git.openjdk.org/jdk/pull/12328.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12328#issuecomment-1410220195)

### CODE REVIEW

**Comment by stsypanov** [comment]
Btw, in `ResourceBundle.CacheKey.equals()` can we replace
```java
Module module = getModule();
Module caller = getCallerModule();
return Objects.equals(module, otherEntry.getModule()) &&
       Objects.equals(caller, otherEntry.getCallerModule());
```
with
```java
return moduleRef.refersTo(otherEntry.getModule()) &&
        callerRef.refersTo(otherEntry.getCallerModule());
```
?

**Comment by mlchung** [comment]
```
return moduleRef.refersTo(otherEntry.getModule()) &&
        callerRef.refersTo(otherEntry.getCallerModule());
```

You need to check if module and caller module are non-null. 

`equals` returns true only if the module and callerModule of this `CacheKey` and other `CacheKey` are the same and non-null.

**Comment by stsypanov** [comment]
Let's wait

**Comment by stsypanov** [comment]
Let's wait

**Comment by stsypanov** [comment]
Keep alive

**Comment by stsypanov** [comment]
Not now

**Comment by stsypanov** [comment]
Any review?

File: src/java.base/share/classes/java/util/ResourceBundle.java [hunk 1]
```diff
@@ -721,26 +721,24 @@ public boolean equals(Object other) {
...
-                if (!locale.equals(otherEntry.locale)) {
+                if (!Objects.equals(locale, otherEntry.locale)) {
                     return false;
                 }
                 // are modules and callerModules the same and non-null?
                 Module module = getModule();
                 Module caller = getCallerModule();
                 return ((module != null) && (module.equals(otherEntry.getModule())) &&
                         (caller != null) && (caller.equals(otherEntry.getCallerModule())));
-            } catch (NullPointerException | ClassCastException e) {
```
Comment (pavelrappo) [inline src/java.base/share/classes/java/util/ResourceBundle.java:743]: Are we sure that NPE can only be thrown when calling equals on `name` or `locale` and not, for example, from `getModule()` or `getCallerModule()`?

**Review by pavelrappo — COMMENTED** [review]
Coincidentally, I'm looking at modernizing some of the equals methods in java.base.

File: src/java.base/share/classes/java/util/ResourceBundle.java (hunk 1 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/util/ResourceBundle.java:743]: Yes:
```java
CacheKey(String baseName, Locale locale, Module module, Module caller) {
    Objects.requireNonNull(module);
    Objects.requireNonNull(caller);

    this.name = baseName;
    this.locale = locale;
    this.moduleRef = new KeyElementReference<>(module, referenceQueue, this);
    this.callerRef = new KeyElementReference<>(caller, referenceQueue, this);
    this.modulesHash = module.hashCode() ^ caller.hashCode();
}

Comment (pavelrappo) [reply src/java.base/share/classes/java/util/ResourceBundle.java:743]: I'll run our CI, and if all good, I'll approve this PR. If nothing else, this change seems reasonable and correct.

Comment (naotoj) [reply src/java.base/share/classes/java/util/ResourceBundle.java:743]: Thanks Pavel for jumping in. I too think this change looks good.

Comment (pavelrappo) [reply src/java.base/share/classes/java/util/ResourceBundle.java:743]: Yep; all good.

**Review by pavelrappo — APPROVED** [review]
Looks good; thanks.

**Review by naotoj — APPROVED** [review]

File: src/java.base/share/classes/sun/util/resources/Bundles.java [hunk 2]
```diff
@@ -502,15 +502,13 @@ public boolean equals(Object other) {
...
-                final CacheKey otherEntry = (CacheKey)other;
+            if (other instanceof CacheKey otherEntry) {
                 //quick check to see if they are not equal
                 if (hashCodeCache != otherEntry.hashCodeCache) {
                     return false;
                 }
-                return locale.equals(otherEntry.locale)
-                        && name.equals(otherEntry.name);
-            } catch (NullPointerException | ClassCastException e) {
+                return Objects.equals(locale, otherEntry.locale)
```
Comment (pavelrappo) [inline src/java.base/share/classes/sun/util/resources/Bundles.java:510]: While the proposal is not wrong, you might want to consider reverting the change to those two lines. The reason is that neither `locale` nor `name` can be null, if I understand this class correctly. 

NPE was seemingly caught to cover for `other` being null.

Comment (pavelrappo) [reply src/java.base/share/classes/sun/util/resources/Bundles.java:510]: > NPE was seemingly caught to cover for other being null.
Unlike that of other file in this PR.

Comment (stsypanov) [reply src/java.base/share/classes/sun/util/resources/Bundles.java]: Done

**Comment by pavelrappo** [comment]
I'll re-run our CI, and if all good, I'll sponsor this PR.

**Comment by pavelrappo** [comment]
> I'll re-run our CI, and if all good, I'll sponsor this PR.

The CI tests I started have just passed. While this PR is already good, I wonder if we make it even better.

I doubt highly that we need null-checks for CacheKey's name and locale. My suspicion comes from these two facts:

* CacheKey's hashCode calls methods on these fields without checking them for null, and
* CacheKey instances are queried and stored in ConcurrentHashMap, which certainly uses both `equals` and `hashCode`

So, what I'm saying is that if those fields were null, there would be unhandled NPE somewhere; but I cannot see such NPE.  Additionally, those fields are used in toString(). Furthermore, locale is known to not be null because the only place from where it is passed to a CacheKey consturctor is ResourceBundle.getBundleImpl(Module, Module, String, Locale, ResourceBundle.Control), which explicitly checks that locale for not being null.

Does it make sense?

**Comment by stsypanov** [comment]
@pavelrappo you mean we can revert changes to ResourceBundle.java in lines 729 and 733?

**Comment by pavelrappo** [comment]
> @pavelrappo you mean we can revert changes to ResourceBundle.java in lines 729 and 733?

Yes, those lines. We can revert to x.equals(otherEntry.x).

**Comment by pavelrappo** [comment]
Thanks, the fix now looks good and simple; testing.

**Review by pavelrappo — APPROVED** [review]
Thanks, the fix now looks good and simple; testing.

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**